src/main/java/org/manifold/compiler/front/ExpressionGraphBuilder.java

```diff
@@ -0,0 +1,201 @@
+      createVariablesFromScope(nsID, ns.getPrivateScope());
+    }
+
+    for (Expression expr : expressions) {
+      expr.accept(this);
+    }
+
+    return exprGraph;
+  }
+
+  private void createVariablesFromScope(NamespaceIdentifier nsID,
+      Scope scope) throws MultipleDefinitionException {
+    for (VariableIdentifier vID : scope.getSymbolIdentifiers()) {
+      // scopes don't know their own namespaces, which is why we need nsID here
+      VariableIdentifier qualifiedID = new VariableIdentifier(
+          nsID, vID.getName());
+      log.debug("constructing vertex for '" + qualifiedID + "'");
+      exprGraph.createVariableVertex(qualifiedID);
+    }
+  }
+
+  @Override
+  public void visit(FunctionInvocationExpression fExpr) {
+    throw new UndefinedBehaviourError("don't know how to visit "
+        + "function invocation expression");
+  }
+
+  @Override
+  public void visit(LiteralExpression lExpr) {
+    ConstantValueVertex v = new ConstantValueVertex(exprGraph,
+        lExpr.getValue());
+    exprGraph.addNonVariableVertex(v);
+    this.lastVertex = v;
+  }
+
+  @Override
+  public void visit(TupleTypeValueExpression tExpr) throws Exception {
+    Map<String, ExpressionEdge> typeValueEdges = new HashMap<>();
+    for (Map.Entry<String, Expression> e 
+        : tExpr.getTypeValueExpressions().entrySet()) {
+      String identifier = e.getKey();
+      e.getValue().accept(this);
+      ExpressionEdge eTypeValue = new ExpressionEdge(lastVertex, null);
+      typeValueEdges.put(identifier, eTypeValue);
+      exprGraph.addEdge(eTypeValue);
+    }
+    Map<String, ExpressionEdge> defaultValueEdges = new HashMap<>();
+    for (Map.Entry<String, Expression> e 
+        : tExpr.getDefaultValueExpressions().entrySet()) {
+      String identifier = e.getKey();
+      e.getValue().accept(this);
+      ExpressionEdge eDefaultValue = new ExpressionEdge(lastVertex, null);
+      defaultValueEdges.put(identifier, eDefaultValue);
+      exprGraph.addEdge(eDefaultValue);
+    }
+    TupleTypeValueVertex vTuple = new TupleTypeValueVertex(exprGraph,
+        typeValueEdges, defaultValueEdges);
+    exprGraph.addNonVariableVertex(vTuple);
+    this.lastVertex = vTuple;
+  }
+  
+  @Override
+  public void visit(FunctionTypeValueExpression fExpr) throws Exception {
+    // get the vertex corresponding to the input type
+    fExpr.getInputTypeExpression().accept(this);
+    ExpressionVertex vIn = lastVertex;
+    ExpressionEdge eIn = new ExpressionEdge(vIn, null);
+    // then get the output type vertex
+    fExpr.getOutputTypeExpression().accept(this);
+    ExpressionVertex vOut = lastVertex;
+    ExpressionEdge eOut = new ExpressionEdge(vOut, null);
+    
+    FunctionTypeValueVertex vFunctionType = new FunctionTypeValueVertex(
+        exprGraph, eIn, eOut);
+    exprGraph.addNonVariableVertex(vFunctionType);
+    exprGraph.addEdge(eIn);
+    exprGraph.addEdge(eOut);
+    this.lastVertex =  vFunctionType;
```

Comment:
Honestly after what I've seen of checkstyle so far, I'm amazed this wasn't a violation.